cli/jobs/single-step/r/accidents/src/aml_utils.R

```diff
@@ -0,0 +1,92 @@
+# AML utility to extend support for MLflow in Azure Machine Learning (AML). This utility does the following:
+# 1. Understands AML MLflow tracking url by extending OSS MLflow R client.
+# 2. Manages AML Token refresh for remote runs (runs that execute in Azure Machine Learning). It uses tcktk2 R libraray to schedule token refresh.
+#    Token refresh interval can be controlled by setting the environment variable MLFLOW_AML_TOKEN_REFRESH_INTERVAL and defaults to 30 seconds.
+
+library(mlflow)
+library(httr)
+library(later)
+library(tcltk2)
+
+new_mlflow_client.mlflow_azureml <- function(tracking_uri) {
+  host <- paste("https", tracking_uri$path, sep = "://")
+  get_host_creds <- function () {
+    mlflow:::new_mlflow_host_creds(
+      host = host,
+      token = Sys.getenv("MLFLOW_TRACKING_TOKEN"),
+      username = Sys.getenv("MLFLOW_TRACKING_USERNAME", NA),
+      password = Sys.getenv("MLFLOW_TRACKING_PASSWORD", NA),
+      insecure = Sys.getenv("MLFLOW_TRACKING_INSECURE", NA)
+    )
+  }
+  cli_env <- function() {
+    creds <- get_host_creds()
+    res <- list(
+      MLFLOW_TRACKING_USERNAME = creds$username,
+      MLFLOW_TRACKING_PASSWORD = creds$password,
+      MLFLOW_TRACKING_TOKEN = creds$token,
+      MLFLOW_TRACKING_INSECURE = creds$insecure
+    )
+    res[!is.na(res)]
+  }
+  mlflow:::new_mlflow_client_impl(get_host_creds, cli_env, class = "mlflow_azureml_client")
+}
+
+get_auth_header <- function() {
+    headers <- list()
+    auth_token <- Sys.getenv("MLFLOW_TRACKING_TOKEN")
+    auth_header <- paste("Bearer", auth_token, sep = " ")
+    headers$Authorization <- auth_header
+    headers
+}
+
+get_token <- function(host, exp_id, run_id) {
+    req_headers <- do.call(httr::add_headers, get_auth_header())
+    token_host <- gsub("mlflow/v1.0","history/v1.0", host)
+    token_host <- gsub("azureml://","https://", token_host)
+    api_url <- paste0(token_host, "/experimentids/", exp_id, "/runs/", run_id, "/token")
+    GET( api_url, timeout(getOption("mlflow.rest.timeout", 60)), req_headers)
+}
+
+
+fetch_token_from_aml <- function() {
+    print("Refreshing token")
+    tracking_uri <- Sys.getenv("MLFLOW_TRACKING_URI")
+    exp_id <- Sys.getenv("MLFLOW_EXPERIMENT_ID")
```

Comment:
not that it matters but you shouldn't need expid

Reply:
For getting a token yes. API signature needs it but backend service does not care about it.